hushline/premium.py

```diff
@@ -0,0 +1,536 @@
+
+
+def handle_invoice_created(invoice: stripe.Invoice) -> None:
+    # invoice.created
+
+    try:
+        new_invoice = StripeInvoice(invoice)
+        db.session.add(new_invoice)
+        db.session.commit()
+    except ValueError as e:
+        current_app.logger.error(f"Error creating invoice: {e}")
+
+
+def handle_invoice_updated(invoice: stripe.Invoice) -> None:
+    # invoice.updated
+
+    stripe_invoice = db.session.query(StripeInvoice).filter_by(invoice_id=invoice.id).first()
+    if stripe_invoice:
+        stripe_invoice.total = invoice.total
+        stripe_invoice.status = StripeInvoiceStatusEnum(invoice.status)
+        db.session.commit()
+    else:
+        raise ValueError(f"Could not find invoice with ID {invoice.id}")
+
+
+async def worker(app: Flask) -> None:
+    current_app.logger.error("Starting worker")
+    with app.app_context():
+        while True:
+            stripe_event = (
+                db.session.query(StripeEvent)
+                .filter_by(status="pending")
+                .order_by(StripeEvent.created_at)
+                .first()
+            )
+            if not stripe_event:
+                await asyncio.sleep(10)
+                continue
+
+            stripe_event.status = "in_progress"
+            db.session.add(stripe_event)
+            db.session.commit()
+
+            event_json = json.loads(stripe_event.event_data)
+            event = stripe.Event.construct_from(event_json, current_app.config["STRIPE_SECRET_KEY"])
+
+            current_app.logger.info(
+                f"Processing event {stripe_event.event_type} ({stripe_event.event_id})"
+            )
+            try:
+                # subscription events
+                if event.type.startswith("customer.subscription."):
+                    subscription: stripe.Subscription = stripe.Subscription.construct_from(
+                        event.data.object, current_app.config["STRIPE_SECRET_KEY"]
+                    )
+                    if event.type == "customer.subscription.created":
+                        handle_subscription_created(subscription)
+                    elif event.type == "customer.subscription.updated":
+                        handle_subscription_updated(subscription)
+                    elif event.type == "customer.subscription.deleted":
+                        handle_subscription_deleted(subscription)
+                # invoice events
+                elif event.type.startswith("invoice."):
+                    invoice: stripe.Invoice = stripe.Invoice.construct_from(
+                        event.data.object, current_app.config["STRIPE_SECRET_KEY"]
+                    )
+                    if event.type == "invoice.created":
+                        handle_invoice_created(invoice)
+                    elif event.type == "invoice.updated":
+                        handle_invoice_updated(invoice)
+
+            except Exception as e:
+                current_app.logger.error(
```

Comment:
We probably want the full stack trace and `exc_info=True` gets that for

Reply:
Fixed in ff5f2173ca30559812e5b4a9652b6355b061200a